istioctl/pkg/writer/compare/sds/util.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sdscompare
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"istio.io/istio/istioctl/pkg/util/configdump"
+	"istio.io/istio/security/pkg/nodeagent/sds"
+	"istio.io/pkg/log"
+)
+
+// SecretItemDiff represents a secret that has been diffed between nodeagent and proxy
+type SecretItemDiff struct {
+	Agent string `json:"agent"`
+	Proxy string `json:"proxy"`
+	SecretItem
+}
+
+// SecretItem is an intermediate representation of secrets, used to provide a common
+// format between the envoy proxy secrets and node agent output which can be diffed
+type SecretItem struct {
+	Name        string `json:"resource_name"`
+	Data        string `json:"cert"`
+	Source      string `json:"source"`
+	Destination string `json:"destination"`
+	SecretMeta
+}
+
+// SecretMeta holds selected fields which can be extracted from parsed x509 cert
+type SecretMeta struct {
+	Valid        bool   `json:"cert_valid"`
+	SerialNumber string `json:"serial_number"`
+	Expires      string `json:"expires"`
+	ValidFrom    string `json:"valid_from"`
+	Type         string `json:"type"`
```

Comment:
I forgot to check. Do you implement the validation context type?

I mean `validation_context` typed secrets. That sometimes can also be the source of the bug? Appears to me we only have key/cert typed secrets.

Reply:
The feature includes CA secrets, with cert data taken from the `trusted_ca` sub-field of `validation_context`. The feature does _not_ diff private keys, since they are redacted on the config dump and omitted from the node agent `/debug`.

Reply:
okay, that's one use case of validation context, but that's only usage I'm aware of Istio's usage of validation context on SDS. So I think that's good enough.

We don't override SAN verification stuff fby SDS, so I think it's fine, at least for now.

/cc @JimmyCYJ can confirm.